plugin-server/src/main/ingestion-queues/session-recording/services/partition-locker.ts

```diff
@@ -0,0 +1,134 @@
+        return [...keys]
+    }
+    /* 
+        Claim the lock for partitions for this consumer
+        - If already locked, we extend the TTL
+        - If it is claimed, we wait and retry until it is cleared 
+        - If unclaimed, we claim it
+    */
+    public async claim(tps: TopicPartition[]) {
+        const keys = this.keys(tps)
+        const blockingConsumers = new Set(...[this.consumerID])
+
+        try {
+            while (blockingConsumers.size !== 0) {
+                blockingConsumers.clear()
+
+                await this.run(`claim keys that belong to this consumer`, async (client) => {
+                    await Promise.allSettled(
+                        keys.map(async (key) => {
+                            const existingClaim = await client.get(key)
+
+                            if (existingClaim && existingClaim !== this.consumerID) {
+                                // Still claimed by someone else!
+                                blockingConsumers.add(existingClaim)
+                                return
+                            }
+
+                            // Set the key so it is claimed by us
+                            const res = await client.set(key, this.consumerID, 'PX', this.ttl)
+                            if (!res) {
+                                blockingConsumers.add(this.consumerID)
+                            }
+                        })
+                    )
+                })
+
+                if (blockingConsumers.size > 0) {
+                    status.warn(
+                        '🔒',
+                        `PartitionLocker failed to claim keys. Waiting ${this.delay} before retrying...`,
+                        {
+                            id: this.consumerID,
+                            blockingConsumers,
+                        }
+                    )
+                    await new Promise((r) => setTimeout(r, this.delay))
+                }
+            }
+
+            status.info('🔒', 'PartitionLocker claimed all required keys')
+        } catch (error) {
+            status.error('🧨', 'PartitionLocker errored to claim keys', {
+                error: error.message,
+                keys,
+            })
+            captureException(error, {
+                extra: {
+                    keys,
+                },
+            })
+        }
+    }
+
+    /* 
+        Release a lock for a partition
+        - Clear our claim if it is set to our consumer so that another can claim it
+    */
+    public async release(tps: TopicPartition[]) {
+        const keys = this.keys(tps)
+        try {
+            await this.run(`release keys that belong to this consumer`, async (client) => {
+                await Promise.allSettled(
+                    keys.map(async (key) => {
+                        const value = await client.get(key)
+                        if (value === this.consumerID) {
```

Comment:
makes me wonder if the lock value should be {consumerId, timeLockTaken} so we can log "tried to release a lock but it was taken by consumer X at time Y"

Might just be noise though... if this never happens then we don't care 🤷

Reply:
Thats sort of what the TTL is for though. We could query the current TTL in the future